### DISCUSSION TITLE
Save chat context on database

### DISCUSSION THREAD

**Discussion by cisco0x60** [telegraf/telegraf#1942] [Q&A] [answered]
I need to create a queue of all clients who perform a certain action with my bot.
This queue need to be persistent and thus, I thought to store them in a database.

However, how can I store the entire context on a database? I need the context since I need to call the context functions (i.e. .reply(), .deleteMessage(), .wizard.next(), etc...).

I attach an (hopefully) quite explanatory image of the problem. 
![image](https://github.com/telegraf/telegraf/assets/32491001/8aa8d9ae-7771-4d68-b699-b579d811601b)

How can I handle this?
Thank you

**Comment by MKRhere** [comment, marked as answer]
You can safely store the contents of `ctx.update` which will always be a valid JSON as received from Telegram. You can construct it back into a context by doing `new Context(update)`.

```TS
const update = ctx.update;
// store update object in DB

// later, retrieve it back
const update = await fetchUpdateFromDbSomehow();
const ctx = new Context(update);
```

**Reply by cisco0x60** [reply]
Oh ok thank you so much, it was exactly what I was looking for!

**Reply by watate** [reply]
Hey @francestu96 was wondering how you were able to do this! 

I tried reconstructing ctx from both ctx.update and ctx itself but I can't use any methods on the reconstructed ctx (ctx.reply doesn't work for example).

To store ctx and ctx.json, at first I did JSON.stringify but because ctx is a Circular JSON it didn't work so then I tried flatten.stringify which allowed me to get back ctx or reconstruct a new ctx. But ctx.reply didn't work.

However, when I try to reconstruct ctx without having it go through flatted.stringify and flagged.parse, ctx.reply works perfectly.

**Reply by MKRhere** [reply]
You should store `ctx.update` (it's a plain object; stringify if needed), and parse it back before passing to `new Context`.

**Reply by watate** [reply]
Hi @MKRhere thanks for the reply! I did JSON.stringify => JSON.parse and also tried flatted.stringify => flatted.parse but ctx.reply didn't work for me. I'll try to see if I can write something to show my code when I'm back in 2-3 weeks! Thanks again.
